### ISSUE TITLE
function return type checking fails to check promise inside of a promise (angular)

### ISSUE THREAD

**Issue by yklee1013** [Microsoft/TypeScript#7057]
eg: 
```
function test() : ng.IPromise<number> {
   var p : ng.IPromise<number> = querySomeApi();

    return p.then(function(val : number)){
     // suppose $q was injected before   
     return $q.when(val);
    })
}
```

This would pass the type checking but it shouldn't. promise.then() will return a promise that wraps around the returned value(val). So this function is actually returning ng.IPromise\<ng.IPromise\<number\>\>.

using ts 1.7.5

**Comment by DanielRosenwasser** [comment]
Hi @yklee1013

If we're talking about [the same `IPromise`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/bcd5761826eb567876c197ccc6a87c4d05731054/angularjs/angular.d.ts#L1065-L1083), then this is behaving correctly. `then` takes a callback that returns either a promise of some type (`IPromise<TResult>`) or some type (`TResult`). `then` will return a promise of the type in focus (`IPromise<TResult>). This flattening is the way [A+ promises](https://promisesaplus.com/) are supposed to work.

**Comment by yklee1013** [comment]
Thanks for reply @DanielRosenwasser ! 

I wasn't specific enough when I posted the issue, but this happens when I was doing some testing for an angular app. I had to manually trigger digest cycles. The digest cycle wasn't triggered for $q.when(val) at the time the control reaches that line, so it was still an IPromise not being resolved. 

The official angular doc for then():  
`This method returns a new promise which is resolved or rejected via the return value of the successCallback, errorCallback (unless that value is a promise, in which case it is resolved with the value which is resolved in that promise using promise chaining). `
[https://docs.angularjs.org/api/ng/service/$q](url)

Since the innermost promise hasn't been resolved, then() returns a promise wrapping another promise that was in the process of getting resolved. Inspecting on console I saw the function returning an IPromise whose $$state.status = -1, which means that it has another IPromise inside. This is really an edge case that I have only encountered while doing testing. Thanks!

**Comment by HerringtonDarkholme** [comment]
Hi @yklee1013 

I think you may misunderstand the meaning of `Promise<T>` here. `Promise` means the value **will be** resolved **asynchronously**, that is, your `test` function should always return an IPromise whose status is pending, when you inspect it synchronously.  It will be settled *after a while*, loosely speaking. Strictly speaking, the promise will be resolved asynchronously, after your `querySomeApi` is resolved.

```typescript
var promise = test()
console.log(promise.$$status.status) // should be pending here
setTimeout(() => console.log(promise.$$status.status), 1000) // may be resolved here, if querySomeApi is resolved before.
```

`Promise<T>` has already abstracted the notion of **deferred value**: your promise will not be resolved immediately after you created it, it will be resolved after the underlying work has been done, asynchronously. There is no difference between a resolved promise and a pending one at type level, in the sense that they have the same type: `Promise<T>`.

Accordingly, there is little point to make a type of `Promise<Promise<T>>`, since type information does not indicate whether a promise is resolved. `Promise<Promise<T>>` means a deferred value of type `T`, same as `Promise<T>`. Whether a promise is resolved depends on runtime execution.

If you still have some confusion, [stackoverflow](http://stackoverflow.com/) is your best friend! Also [Q's wiki page](https://github.com/kriskowal/q/wiki/General-Promise-Resources) may also help
